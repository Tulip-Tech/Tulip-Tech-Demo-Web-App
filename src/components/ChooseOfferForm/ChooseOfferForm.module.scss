@use 'src/styles/variables';
@use 'src/styles/theme';
@use 'src/styles/mixins/responsive';

.title {
  margin-bottom: 8px;
  font-weight: var(--body-font-weight-bold);
  font-size: 24px;
}

.subtitle {
  margin-bottom: 24px;
  font-weight: var(--body-font-weight-bold);
  font-size: 18px;
}

.offerGroupSwitch {
  display: flex;
  flex: 1;
  justify-content: center;
  margin-bottom: 24px;

  > input {
    &:checked + .label {
      border-radius: 4px;
      transform: scale(1.02);
    }
  }
}

.offers {
  display: flex;
  margin: 0 -4px 24px;
}

.offer {
  flex: 1;
  margin: 0 4px;
}

.radio {
  position: absolute;
  width: 1px;
  height: 1px;
  overflow: hidden;
  white-space: nowrap;
  clip: rect(0 0 0 0);
  clip-path: inset(50%);

  :focus,
  :active {
    + .label {
      border-color: variables.$white;
    }
  }

  &:checked + .label {
    color: variables.$black;
    background-color: variables.$white;
    border-color: variables.$white;
  }
}

.label {
  display: flex;
  flex-direction: column;
  height: 100%;
  padding: 16px;
  background-color: rgba(variables.$black, 0.34);
  border: 1px solid rgba(variables.$white, 0.34);
  border-radius: 4px;
  cursor: pointer;
  transition: border 0.2s ease, background 0.2s ease, transform 0.3s ease-out;
}

.offerGroupLabel {
  padding: 8px 16px;
  font-weight: var(--body-font-weight-bold);
  font-size: 20px;

  &:nth-child(2) {
    border-radius: 4px 0 0 4px;
  }
  &:nth-child(4) {
    border-radius: 0 4px 4px 0;
  }
}

.offerTitle {
  font-weight: var(--body-font-weight-bold);
  font-size: 20px;
  text-align: center;
}

.offerDivider {
  width: 100%;
  border: none;
  border-bottom: 1px solid currentColor;
  opacity: 0.54;
}

.offerBenefits {
  margin-bottom: 16px;
  padding: 0;

  > li {
    display: flex;
    align-items: center;
    margin-bottom: 4px;
    padding: 4px 0;

    > svg {
      flex-shrink: 0;
      margin-right: 4px;
      fill: variables.$green;
    }

    @include responsive.mobile-only() {
      font-size: 14px;
    }
  }
}

.fill {
  flex: 1;
}

.offerPrice {
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  align-items: baseline;
  font-size: 32px;

  > small {
    margin-left: 4px;
    font-size: 12px;
  }
}
