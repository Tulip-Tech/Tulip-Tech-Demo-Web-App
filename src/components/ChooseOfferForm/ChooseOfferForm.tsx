import React from 'react';
import { useTranslation } from 'react-i18next';
import classNames from 'classnames';

import styles from './ChooseOfferForm.module.scss';

import Button from '#components/Button/Button';
import FormFeedback from '#components/FormFeedback/FormFeedback';
import DialogBackButton from '#components/DialogBackButton/DialogBackButton';
import LoadingOverlay from '#components/LoadingOverlay/LoadingOverlay';
import CheckCircle from '#src/icons/CheckCircle';
import type { Offer } from '#types/checkout';
import { getOfferPrice, isSVODOffer } from '#src/utils/subscription';
import type { FormErrors } from '#types/form';
import { testId } from '#src/utils/common';
import type { ChooseOfferFormData, OfferType } from '#types/account';
import { useConfigStore } from '#src/stores/ConfigStore';

type Props = {
  values: ChooseOfferFormData;
  errors: FormErrors<ChooseOfferFormData>;
  onChange: React.ChangeEventHandler<HTMLInputElement | HTMLTextAreaElement>;
  onSubmit: React.FormEventHandler<HTMLFormElement>;
  onBackButtonClickHandler?: () => void;
  offers: Offer[];
  submitting: boolean;
  offerType: OfferType;
  setOfferType?: (offerType: OfferType) => void;
};

type OfferBoxProps = {
  offer: Offer;
  title: string;
  ariaLabel: string;
  secondBenefit?: string;
  periodString?: string;
};

type OfferBox = ({ offer, title, ariaLabel, secondBenefit, periodString }: OfferBoxProps) => JSX.Element;

const ChooseOfferForm: React.FC<Props> = ({
  values,
  errors,
  onChange,
  onSubmit,
  submitting,
  offers,
  onBackButtonClickHandler,
  offerType,
  setOfferType,
}: Props) => {
  const siteName = useConfigStore((s) => s.config.siteName);
  const { t } = useTranslation('account');

  const getFreeTrialText = (offer: Offer) => {
    if (offer.freeDays) {
      return t('choose_offer.benefits.first_days_free', { count: offer.freeDays });
    } else if (offer.freePeriods) {
      // t('periods.day')
      // t('periods.week')
      // t('periods.month')
      // t('periods.year')
      const period = t(`periods.${offer.period}`, { count: offer.freePeriods });

      return t('choose_offer.benefits.first_periods_free', { count: offer.freePeriods, period });
    }

    return null;
  };

  const OfferBox: OfferBox = ({ offer, title, ariaLabel, secondBenefit, periodString }) => (
    <div className={styles.offer}>
      <input
        className={styles.radio}
        onChange={onChange}
        type="radio"
        name={'offerId'}
        value={offer.offerId}
        id={offer.offerId}
        checked={values.offerId === offer.offerId}
        aria-label={ariaLabel}
      />
      <label className={styles.label} htmlFor={offer.offerId}>
        <h4 className={styles.offerTitle}>{title}</h4>
        <hr className={styles.offerDivider} />
        <ul className={styles.offerBenefits}>
          {offer.freeDays || offer.freePeriods ? (
            <li>
              <CheckCircle /> {getFreeTrialText(offer)}
            </li>
          ) : null}

          {!!secondBenefit && (
            <li>
              <CheckCircle /> {secondBenefit}
            </li>
          )}
          <li>
            <CheckCircle /> {t('choose_offer.benefits.watch_on_all_devices')}
          </li>
        </ul>
        <div className={styles.fill} />
        <div className={styles.offerPrice}>
          {getOfferPrice(offer)} {!!periodString && <small>/{periodString}</small>}
        </div>
      </label>
    </div>
  );

  const renderOfferBox = (offer: Offer) => {
    if (isSVODOffer(offer)) {
      const isMonthly = offer.period === 'month';

      return (
        <OfferBox
          offer={offer}
          key={offer.offerId}
          title={isMonthly ? t('choose_offer.monthly') : t('choose_offer.yearly')}
          ariaLabel={isMonthly ? t('choose_offer.monthly_subscription') : t('choose_offer.yearly_subscription')}
          secondBenefit={t('choose_offer.benefits.cancel_anytime')}
          periodString={isMonthly ? t('periods.month') : t('periods.year')}
        />
      );
    }

    return (
      <OfferBox
        offer={offer}
        key={offer.offerId}
        title={offer.offerTitle}
        ariaLabel={offer.offerTitle}
        secondBenefit={
          !!offer.durationPeriod && !!offer.durationAmount
            ? t('choose_offer.tvod_access', { period: offer.durationPeriod, count: offer.durationAmount })
            : undefined
        }
      />
    );
  };

  return (
    <form onSubmit={onSubmit} data-testid={testId('choose-offer-form')} noValidate>
      {onBackButtonClickHandler ? <DialogBackButton onClick={onBackButtonClickHandler} /> : null}
      <h2 className={styles.title}>{t('choose_offer.title')}</h2>
      <h3 className={styles.subtitle}>{t('choose_offer.watch_this_on_platform', { siteName })}</h3>
      {errors.form ? <FormFeedback variant="error">{errors.form}</FormFeedback> : null}
      {setOfferType && (
        <div className={styles.offerGroupSwitch}>
          <input
            className={styles.radio}
            onChange={() => setOfferType('svod')}
            type="radio"
            name="offerType"
            id="svod"
            value="svod"
            checked={offerType === 'svod'}
          />
          <label className={classNames(styles.label, styles.offerGroupLabel)} htmlFor="svod">
            {t('choose_offer.subscription')}
          </label>
          <input
            className={styles.radio}
            onChange={() => setOfferType('tvod')}
            type="radio"
            name="offerType"
            id="tvod"
            value="tvod"
            checked={offerType === 'tvod'}
          />
          <label className={classNames(styles.label, styles.offerGroupLabel)} htmlFor="tvod">
            {t('choose_offer.one_time_only')}
          </label>
        </div>
      )}
      <div className={styles.offers}>{offers.map(renderOfferBox)}</div>
      {submitting && <LoadingOverlay transparentBackground inline />}
      <Button label={t('choose_offer.continue')} disabled={submitting} variant="contained" color="primary" type="submit" fullWidth />
    </form>
  );
};
export default ChooseOfferForm;
