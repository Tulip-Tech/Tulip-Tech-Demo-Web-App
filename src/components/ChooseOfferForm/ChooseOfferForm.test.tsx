import React from 'react';
import { fireEvent, render } from '@testing-library/react';

import ChooseOfferForm from './ChooseOfferForm';

import monthlyOffer from '#test/fixtures/monthlyOffer.json';
import yearlyOffer from '#test/fixtures/yearlyOffer.json';
import tvodOffer from '#test/fixtures/tvodOffer.json';
import type { Offer } from '#types/checkout';

const svodOffers = [monthlyOffer, yearlyOffer] as Offer[];
const tvodOffers = [tvodOffer] as Offer[];

describe('<OffersForm>', () => {
  test('renders and matches snapshot', () => {
    const { container } = render(
      <ChooseOfferForm
        values={{ offerId: 'S916977979_NL' }}
        errors={{}}
        onChange={vi.fn()}
        onSubmit={vi.fn()}
        submitting={false}
        offers={svodOffers}
        offerType={'svod'}
        setOfferType={vi.fn()}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  test('renders and matches snapshot', () => {
    const { container } = render(
      <ChooseOfferForm
        values={{ offerId: 'R892134629_NL' }}
        errors={{}}
        onChange={vi.fn()}
        onSubmit={vi.fn()}
        submitting={false}
        offers={tvodOffers}
        offerType={'tvod'}
        setOfferType={vi.fn()}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  test('checks the monthly offer correctly', () => {
    const { getByLabelText } = render(
      <ChooseOfferForm
        values={{ offerId: 'S916977979_NL' }}
        errors={{}}
        onChange={vi.fn()}
        onSubmit={vi.fn()}
        submitting={false}
        offers={svodOffers}
        offerType={'svod'}
        setOfferType={vi.fn()}
      />,
    );

    expect(getByLabelText('choose_offer.monthly_subscription')).toBeChecked();
  });

  test('checks the yearly offer correctly', () => {
    const { getByLabelText } = render(
      <ChooseOfferForm
        values={{ offerId: 'S345569153_NL' }}
        errors={{}}
        onChange={vi.fn()}
        onSubmit={vi.fn()}
        submitting={false}
        offers={svodOffers}
        offerType={'svod'}
        setOfferType={vi.fn()}
      />,
    );

    expect(getByLabelText('choose_offer.yearly_subscription')).toBeChecked();
  });

  test('checks the tvod offer correctly', () => {
    const { getByLabelText } = render(
      <ChooseOfferForm
        values={{ offerId: 'R892134629_NL' }}
        errors={{}}
        onChange={vi.fn()}
        onSubmit={vi.fn()}
        submitting={false}
        offers={tvodOffers}
        offerType={'tvod'}
        setOfferType={vi.fn()}
      />,
    );

    expect(getByLabelText('One Time - TVOD offer')).toBeChecked();
  });

  test('calls the onChange callback when changing the offer', () => {
    const onChange = vi.fn();
    const { getByLabelText } = render(
      <ChooseOfferForm
        values={{ offerId: 'S916977979_NL' }}
        errors={{}}
        onChange={onChange}
        onSubmit={vi.fn()}
        submitting={false}
        offers={svodOffers}
        offerType={'svod'}
        setOfferType={vi.fn()}
      />,
    );

    fireEvent.click(getByLabelText('choose_offer.yearly_subscription'));

    expect(onChange).toBeCalled();
  });

  test('calls the onSubmit callback when submitting the form', () => {
    const onSubmit = vi.fn();
    const { getByTestId } = render(
      <ChooseOfferForm
        values={{ offerId: 'S916977979_NL' }}
        errors={{}}
        onChange={vi.fn()}
        onSubmit={onSubmit}
        submitting={false}
        offers={svodOffers}
        offerType={'svod'}
        setOfferType={vi.fn()}
      />,
    );

    fireEvent.submit(getByTestId('choose-offer-form'));

    expect(onSubmit).toBeCalled();
  });
});
