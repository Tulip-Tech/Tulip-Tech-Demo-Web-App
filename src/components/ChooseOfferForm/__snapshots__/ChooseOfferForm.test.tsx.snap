// Vitest Snapshot v1

exports[`<OffersForm> > renders and matches snapshot 1`] = `
<div>
  <form
    data-testid="choose-offer-form"
    novalidate=""
  >
    <h2
      class="_title_f7961f"
    >
      choose_offer.title
    </h2>
    <h3
      class="_subtitle_f7961f"
    >
      choose_offer.watch_this_on_platform
    </h3>
    <div
      class="_offerGroupSwitch_f7961f"
    >
      <input
        checked=""
        class="_radio_f7961f"
        id="svod"
        name="offerType"
        type="radio"
        value="svod"
      />
      <label
        class="_label_f7961f _offerGroupLabel_f7961f"
        for="svod"
      >
        choose_offer.subscription
      </label>
      <input
        class="_radio_f7961f"
        id="tvod"
        name="offerType"
        type="radio"
        value="tvod"
      />
      <label
        class="_label_f7961f _offerGroupLabel_f7961f"
        for="tvod"
      >
        choose_offer.one_time_only
      </label>
    </div>
    <div
      class="_offers_f7961f"
    >
      <div
        class="_offer_f7961f"
      >
        <input
          aria-label="choose_offer.monthly_subscription"
          checked=""
          class="_radio_f7961f"
          id="S916977979_NL"
          name="offerId"
          type="radio"
          value="S916977979_NL"
        />
        <label
          class="_label_f7961f"
          for="S916977979_NL"
        >
          <h4
            class="_offerTitle_f7961f"
          >
            choose_offer.monthly
          </h4>
          <hr
            class="_offerDivider_f7961f"
          />
          <ul
            class="_offerBenefits_f7961f"
          >
            <li>
              <svg
                aria-hidden="true"
                class="_icon_1e9999"
                focusable="false"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g>
                  <path
                    d="M12 2C6.5 2 2 6.5 2 12S6.5 22 12 22 22 17.5 22 12 17.5 2 12 2M10 17L5 12L6.41 10.59L10 14.17L17.59 6.58L19 8L10 17Z"
                  />
                </g>
              </svg>
               
              choose_offer.benefits.first_periods_free
            </li>
            <li>
              <svg
                aria-hidden="true"
                class="_icon_1e9999"
                focusable="false"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g>
                  <path
                    d="M12 2C6.5 2 2 6.5 2 12S6.5 22 12 22 22 17.5 22 12 17.5 2 12 2M10 17L5 12L6.41 10.59L10 14.17L17.59 6.58L19 8L10 17Z"
                  />
                </g>
              </svg>
               
              choose_offer.benefits.cancel_anytime
            </li>
            <li>
              <svg
                aria-hidden="true"
                class="_icon_1e9999"
                focusable="false"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g>
                  <path
                    d="M12 2C6.5 2 2 6.5 2 12S6.5 22 12 22 22 17.5 22 12 17.5 2 12 2M10 17L5 12L6.41 10.59L10 14.17L17.59 6.58L19 8L10 17Z"
                  />
                </g>
              </svg>
               
              choose_offer.benefits.watch_on_all_devices
            </li>
          </ul>
          <div
            class="_fill_f7961f"
          />
          <div
            class="_offerPrice_f7961f"
          >
            € 6,99
             
            <small>
              /
              periods.month
            </small>
          </div>
        </label>
      </div>
      <div
        class="_offer_f7961f"
      >
        <input
          aria-label="choose_offer.yearly_subscription"
          class="_radio_f7961f"
          id="S345569153_NL"
          name="offerId"
          type="radio"
          value="S345569153_NL"
        />
        <label
          class="_label_f7961f"
          for="S345569153_NL"
        >
          <h4
            class="_offerTitle_f7961f"
          >
            choose_offer.yearly
          </h4>
          <hr
            class="_offerDivider_f7961f"
          />
          <ul
            class="_offerBenefits_f7961f"
          >
            <li>
              <svg
                aria-hidden="true"
                class="_icon_1e9999"
                focusable="false"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g>
                  <path
                    d="M12 2C6.5 2 2 6.5 2 12S6.5 22 12 22 22 17.5 22 12 17.5 2 12 2M10 17L5 12L6.41 10.59L10 14.17L17.59 6.58L19 8L10 17Z"
                  />
                </g>
              </svg>
               
              choose_offer.benefits.first_days_free
            </li>
            <li>
              <svg
                aria-hidden="true"
                class="_icon_1e9999"
                focusable="false"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g>
                  <path
                    d="M12 2C6.5 2 2 6.5 2 12S6.5 22 12 22 22 17.5 22 12 17.5 2 12 2M10 17L5 12L6.41 10.59L10 14.17L17.59 6.58L19 8L10 17Z"
                  />
                </g>
              </svg>
               
              choose_offer.benefits.cancel_anytime
            </li>
            <li>
              <svg
                aria-hidden="true"
                class="_icon_1e9999"
                focusable="false"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g>
                  <path
                    d="M12 2C6.5 2 2 6.5 2 12S6.5 22 12 22 22 17.5 22 12 17.5 2 12 2M10 17L5 12L6.41 10.59L10 14.17L17.59 6.58L19 8L10 17Z"
                  />
                </g>
              </svg>
               
              choose_offer.benefits.watch_on_all_devices
            </li>
          </ul>
          <div
            class="_fill_f7961f"
          />
          <div
            class="_offerPrice_f7961f"
          >
            € 50,00
             
            <small>
              /
              periods.year
            </small>
          </div>
        </label>
      </div>
    </div>
    <button
      aria-disabled="false"
      class="_button_f8f296 _primary_f8f296 _fullWidth_f8f296"
      type="submit"
    >
      <span>
        choose_offer.continue
      </span>
    </button>
  </form>
</div>
`;

exports[`<OffersForm> > renders and matches snapshot 2`] = `
<div>
  <form
    data-testid="choose-offer-form"
    novalidate=""
  >
    <h2
      class="_title_f7961f"
    >
      choose_offer.title
    </h2>
    <h3
      class="_subtitle_f7961f"
    >
      choose_offer.watch_this_on_platform
    </h3>
    <div
      class="_offerGroupSwitch_f7961f"
    >
      <input
        class="_radio_f7961f"
        id="svod"
        name="offerType"
        type="radio"
        value="svod"
      />
      <label
        class="_label_f7961f _offerGroupLabel_f7961f"
        for="svod"
      >
        choose_offer.subscription
      </label>
      <input
        checked=""
        class="_radio_f7961f"
        id="tvod"
        name="offerType"
        type="radio"
        value="tvod"
      />
      <label
        class="_label_f7961f _offerGroupLabel_f7961f"
        for="tvod"
      >
        choose_offer.one_time_only
      </label>
    </div>
    <div
      class="_offers_f7961f"
    >
      <div
        class="_offer_f7961f"
      >
        <input
          aria-label="One Time - TVOD offer"
          checked=""
          class="_radio_f7961f"
          id="R892134629_NL"
          name="offerId"
          type="radio"
          value="R892134629_NL"
        />
        <label
          class="_label_f7961f"
          for="R892134629_NL"
        >
          <h4
            class="_offerTitle_f7961f"
          >
            One Time - TVOD offer
          </h4>
          <hr
            class="_offerDivider_f7961f"
          />
          <ul
            class="_offerBenefits_f7961f"
          >
            <li>
              <svg
                aria-hidden="true"
                class="_icon_1e9999"
                focusable="false"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g>
                  <path
                    d="M12 2C6.5 2 2 6.5 2 12S6.5 22 12 22 22 17.5 22 12 17.5 2 12 2M10 17L5 12L6.41 10.59L10 14.17L17.59 6.58L19 8L10 17Z"
                  />
                </g>
              </svg>
               
              choose_offer.benefits.watch_on_all_devices
            </li>
          </ul>
          <div
            class="_fill_f7961f"
          />
          <div
            class="_offerPrice_f7961f"
          >
            € 2,90
             
          </div>
        </label>
      </div>
    </div>
    <button
      aria-disabled="false"
      class="_button_f8f296 _primary_f8f296 _fullWidth_f8f296"
      type="submit"
    >
      <span>
        choose_offer.continue
      </span>
    </button>
  </form>
</div>
`;
